test/mlir/test/gc/Dialect/Microkernel/linalg-to-microkernel.mlir

```diff
@@ -0,0 +1,175 @@
+// CHECK: %[[B:.+]] = tensor.extract_slice %[[Bsrc:.+]][%[[arg2]], 0, 0, 0] [1, 32, 16, 32] [1, 1, 1, 1] : tensor<8x32x16x32xf32> to tensor<32x16x32xf32>
+// CHECK-NOT: linalg.fill
+// CHECK-NOT: linalg.transpose
+// CHECK: %[[RES:.+]] = microkernel.brgemm ins(%[[A]], %[[B]] : [[TYPE:.+]]) outs(%[[C]] : [[TYPE2:.+]]) batch_dims(0, 1) leading_dims(1, 0) flags(beta_0)  -> tensor<32x32xf32>
+// CHECK-NEXT: scf.forall.in_parallel
+
+// -----
+
+#map1 = affine_map<(d0, d1) -> (d0, d1)>
+func.func @vnni_linalg_to_microkernel_fusing_transpose(%arg0: tensor<4x8x32x32xf32>) -> tensor<4x8x32x32xf32> {
+  %cst = arith.constant 0.000000e+00 : f32
+  %alloc_1 = tensor.empty() : tensor<4x16x32x32xbf16>
+  %alloc_4 = tensor.empty() : tensor<8x16x16x32x2xbf16>
+  %trans_base = tensor.empty() : tensor<16x16x32x2xbf16>
+  %ret = scf.forall (%arg7, %arg8) in (4, 8) shared_outs(%argp = %arg0) -> (tensor<4x8x32x32xf32>) {
+    %alloc_10 = tensor.extract_slice %argp[%arg7, %arg8, 0, 0] [1, 1, 32, 32] [1, 1, 1, 1] : tensor<4x8x32x32xf32> to tensor<32x32xf32>
+    %subview = tensor.extract_slice %alloc_1[%arg7, 0, 0, 0] [1, 16, 32, 32] [1, 1, 1, 1] : tensor<4x16x32x32xbf16> to tensor<16x32x32xbf16>
+    %subview_11 = tensor.extract_slice %alloc_4[%arg8, 0, 0, 0, 0] [1, 16, 16, 32, 2] [1, 1, 1, 1, 1] : tensor<8x16x16x32x2xbf16> to tensor<16x16x32x2xbf16>
+    %11 = linalg.fill ins(%cst : f32) outs(%alloc_10 : tensor<32x32xf32>) -> tensor<32x32xf32>
+    %transposed = linalg.transpose ins(%subview_11 : tensor<16x16x32x2xbf16>) outs(%trans_base : tensor<16x16x32x2xbf16>) permutation = [1, 0, 2, 3]
+    %res = linalgx.batch_reduce_matmul_vnni ins(%subview, %transposed : tensor<16x32x32xbf16>, tensor<16x16x32x2xbf16>) outs(%11 : tensor<32x32xf32>) -> tensor<32x32xf32>
+    scf.forall.in_parallel {
+        tensor.parallel_insert_slice %res into %argp[%arg7, %arg8, 0, 0] [1, 1, 32, 32] [1, 1, 1, 1] : tensor<32x32xf32> into tensor<4x8x32x32xf32>
+    } 
+  }
+  return %ret : tensor<4x8x32x32xf32>
+}
+
+// CHECK-LABEL: vnni_linalg_to_microkernel_fusing_transpose
+// CHECK: scf.forall
+// CHECK: %[[C:.+]] = tensor.extract_slice %[[Csrc:.+]][%[[arg1:.+]], %[[arg2:.+]], 0, 0] [1, 1, 32, 32] [1, 1, 1, 1] : tensor<4x8x32x32xf32> to tensor<32x32xf32>
+// CHECK: %[[A:.+]] = tensor.extract_slice %[[Asrc:.+]][%[[arg1]], 0, 0, 0] [1, 16, 32, 32] [1, 1, 1, 1] : tensor<4x16x32x32xbf16> to tensor<16x32x32xbf16>
+// CHECK: %[[B:.+]] = tensor.extract_slice %[[Bsrc:.+]][%[[arg2]], 0, 0, 0, 0] [1, 16, 16, 32, 2] [1, 1, 1, 1, 1] : tensor<8x16x16x32x2xbf16> to tensor<16x16x32x2xbf16>
+// CHECK-NOT: linalg.fill
+// CHECK-NOT: linalg.transpose
+// CHECK: %[[RES:.+]] = microkernel.brgemm ins(%[[A]], %[[B]] : [[TYPE:.+]]) outs(%[[C]] : [[TYPE2:.+]]) batch_dims(0, 1) leading_dims(1, 0) flags(beta_0)  -> tensor<32x32xf32>
+// CHECK-NEXT: scf.forall.in_parallel
+
+// -----
```

Comment:
It is better to add some tests for the brgemm warped in `scf.if` like
```
func.func @entry(%arg0: memref<4096x4096xf32>, %arg1: memref<128x4096xf32>, %arg2: memref<128x4096xf32>, %arg3: memref<128x4096xf32>) {
    %cst = arith.constant 0.000000e+00 : f32
    %c16 = arith.constant 16 : index
    %c128 = arith.constant 128 : index
    %c1024 = arith.constant 1024 : index
    %c4096 = arith.constant 4096 : index
    %c0 = arith.constant 0 : index
    %0 = bufferization.to_tensor %arg0 restrict : memref<4096x4096xf32>
    %1 = bufferization.to_tensor %arg1 restrict : memref<128x4096xf32>
    %2 = tensor.empty() : tensor<128x4096xf32>
    %3 = tensor.empty() : tensor<64x16x16xf32>
    %4 = scf.forall (%arg4) = (0) to (4096) step (128) shared_outs(%arg5 = %2) -> (tensor<128x4096xf32>) {
      %extracted_slice = tensor.extract_slice %arg5[0, %arg4] [128, 128] [1, 1] : tensor<128x4096xf32> to tensor<128x128xf32>
      %5 = scf.for %arg6 = %c0 to %c4096 step %c1024 iter_args(%arg7 = %extracted_slice) -> (tensor<128x128xf32>) {
        %6 = arith.cmpi eq, %arg6, %c0 : index
        %7 = scf.for %arg8 = %c0 to %c128 step %c16 iter_args(%arg9 = %arg7) -> (tensor<128x128xf32>) {
          %extracted_slice_0 = tensor.extract_slice %1[%arg8, %arg6] [16, 1024] [1, 1] : tensor<128x4096xf32> to tensor<16x1024xf32>
          %extracted_slice_1 = tensor.extract_slice %arg9[%arg8, 0] [16, 128] [1, 1] : tensor<128x128xf32> to tensor<16x128xf32>
          %expanded = tensor.expand_shape %extracted_slice_0 [[0], [1, 2]] output_shape [16, 64, 16] : tensor<16x1024xf32> into tensor<16x64x16xf32>
          %8 = scf.for %arg10 = %c0 to %c128 step %c16 iter_args(%arg11 = %extracted_slice_1) -> (tensor<16x128xf32>) {
            %9 = affine.apply #map()[%arg4, %arg10]
            %extracted_slice_2 = tensor.extract_slice %0[%arg6, %9] [1024, 16] [1, 1] : tensor<4096x4096xf32> to tensor<1024x16xf32>
            %extracted_slice_3 = tensor.extract_slice %arg11[0, %arg10] [16, 16] [1, 1] : tensor<16x128xf32> to tensor<16x16xf32>
            %transposed = linalg.transpose ins(%expanded : tensor<16x64x16xf32>) outs(%3 : tensor<64x16x16xf32>) permutation = [1, 0, 2] 
            %expanded_4 = tensor.expand_shape %extracted_slice_2 [[0, 1], [2]] output_shape [64, 16, 16] : tensor<1024x16xf32> into tensor<64x16x16xf32>
            %10 = scf.if %6 -> (tensor<16x16xf32>) {
              %11 = linalg.fill ins(%cst : f32) outs(%extracted_slice_3 : tensor<16x16xf32>) -> tensor<16x16xf32>
              %12 = linalg.batch_reduce_matmul ins(%transposed, %expanded_4 : tensor<64x16x16xf32>, tensor<64x16x16xf32>) outs(%11 : tensor<16x16xf32>) -> tensor<16x16xf32>
              scf.yield %12 : tensor<16x16xf32>
            } else {
              %11 = linalg.batch_reduce_matmul ins(%transposed, %expanded_4 : tensor<64x16x16xf32>, tensor<64x16x16xf32>) outs(%extracted_slice_3 : tensor<16x16xf32>) -> tensor<16x16xf32>
              scf.yield %11 : tensor<16x16xf32>
            }
            %inserted_slice_5 = tensor.insert_slice %10 into %arg11[0, %arg10] [16, 16] [1, 1] : tensor<16x16xf32> into tensor<16x128xf32>
            scf.yield %inserted_slice_5 : tensor<16x128xf32>
          }
          %inserted_slice = tensor.insert_slice %8 into %arg9[%arg8, 0] [16, 128] [1, 1] : tensor<16x128xf32> into tensor<128x128xf32>
          scf.yield %inserted_slice : tensor<128x128xf32>
        }
        scf.yield %7 : tensor<128x128xf32>
      }
      scf.forall.in_parallel {
        tensor.parallel_insert_slice %5 into %arg5[0, %arg4] [128, 128] [1, 1] : tensor<128x128xf32> into tensor<128x4096xf32>
      }
    }
    bufferization.materialize_in_destination %4 in restrict writable %arg3 : (tensor<128x4096xf32>, memref<128x4096xf32>) -> ()
    return
  }
```

Reply:
test with branch added.